Fixed Flaky Test: CustomDataParametersTest

deekshacheruku opened the pull request:
Fixed the flaky test [CustomDataParametersTest ](https://github.com/deekshacheruku/facebook-java-business-sdk/blob/ac3385534e41db4ab21ffb4acf5f3788e359048c/src/test/java/com/facebook/ads/ServerSideTest.java#L40C1-L100C4) inside src/test/java/com/facebook/ads/ServerSideTest class.


**Root Cause**
The test CustomDataParametersTest has been reported as flaky when ran with the [NonDex](https://github.com/TestingResearchIllinois/NonDex) tool. The test failed because it is trying to check if it contains a string or not which is actually nothing but a String Values of List, Hash Map. The HashMap/List in Java is implemented in such a way that it does not store the order in which the keys and/or values are inserted. As a result, when the expected  string (which is hard-coded) is compared with the actual one, it failed.

**Fix**
The simplest fix is to check if all the patterns of expected String in actual String `serialisedPayload`. Since there were only two possible order changes, it was easy to hardcode the expected patterns and fix the flakiness. 

**How this has been tested?**

*Java*: openjdk version "11.0.20.1"
*Maven*: Apache Maven 3.6.3

*Module build*: Successful
Command used: `mvn install -am -DskipTests`

*Regular test*: Successful
Command used: `mvn test -Dtest=com.facebook.ads.ServerSideTest#CustomDataParametersTest`

*NonDex test*: Failed
Command used: 
`mvn edu.illinois:nondex-maven-plugin:2.1.1:nondex -Dtest=com.facebook.ads.ServerSideTest#CustomDataParametersTest`

NonDex test passed after the fix.

Commit:
Auto-generated java SDK code update

Summary: ignore-conflict-markers

Reviewed By: stcheng

Differential Revision: D50280492

fbshipit-source-id: 0779efafd657f4eac93206c88b717abe01395cdd

Commit:
Fixed Flaky Test: CustomDataParametersTest

FarmersWrap commented:
The fix seems right. But there are a few things I want to comment on.   
1. Is this the only way to fix this test by hardcoding the output? I am good with hard-coded string answers.  
2. It may be clearer if you can put some code analysis in your pull request.

Doom-Prophet commented:
Good fix since there are only two possible order changes for that string, but may get more difficult to fix when the number of element increases. Maybe try with a LinkedHashmap for that scenario and show the idea to the developer will help to get approved.

zzjas commented:
The fix looks ok but an alternative would be to assert the actual contains `key1: value1` and `key2: value2` and then the length of the actual is correct. It's just something for you to consider but it's hard to tell which one the developers will favor more.

Other than that, the text and code look both ok to me. You can proceed to open a real PR. Once you open a real PR, please mark this tentative PR as `Opened` in your `tentative_pr.csv` file and also raise a PR to IDoFT marking this as `Opened`. Thanks!

Commit:
Auto-generated java SDK code update

Summary: ignore-conflict-markers

Reviewed By: stcheng

Differential Revision: D50948084

fbshipit-source-id: 1b663341cd7befb6d4926d254c85ebe234cb6965

stcheng committed:
Update Developer Info

Summary: [BizSDK][Java] Update Developer Info

Reviewed By: ZachYang113

Differential Revision: D51090465

fbshipit-source-id: 0e0f4b09600d0149ff70dbd62469c8241af5eec7
